keps/sig-scheduling/4328-node-affinity-based-eviction/README.md

```diff
@@ -0,0 +1,495 @@
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+Will be tested manually.
+
+Test1:
+- Create a node that has a label
+- Start a local Kubernetes cluster (feature gate defaulted disabled)
+- Create a Pod `test-pod` with `RequiredDuringSchedulingRequiredDuringExecution` node affinity matches the node
+- Check Pod's `requiredDuringSchedulingRequiredDuringExecution` node affinty gets dropped due to disabled feature gate
+- Delete the Pod
+
+- Re-start kube-controller-manager, kube-scheduler and kube-apiserver with feature gate enabled
+- Create `test-pod` again.
+- Verify that the pod is correctly scheduled to the node
+- Remove the node label
+- Verify that the pod is evicted
+- Add the node's label back
+- Create `test-pod` again
+  
+- Re-start kube-controller-manager, kube-scheduler and kube-apiserver with feature gate disabled
+- Verify the old `.spec.affinity.nodeAffinity.requiredDuringSchedulingRequiredDuringExecution` reserved
+- Remove the node label 
+- Verify that the pod is not evicted
+- Add the node's label back 
+
+- Re-start kube-controller-manager, kube-scheduler and kube-apiserver with feature gate enabled
+- Remove the node label
+- Verify that the pod is evicted 
+
+Test2:
+- Start kube-controller-manager, kube-scheduler and kube-apiserver with feature gate enabled.
+- Create a daemonSet with requiredDuringSchedulingRequiredDuringExecution. 
+- Verify that the pod runs only on nodes that meet the nodeAffinity.
+- Remove the node label.
+- Verify that the daemonSet pod on that node is evicted and not recreated.
+- Add the node's label back.
+- Verify that the daemonSet pod runs on the same node again.
+  
+- Re-start kube-controller-manager, kube-scheduler and kube-apiserver with feature gate disabled
+- Verify the old `.spec.affinity.nodeAffinity.requiredDuringSchedulingRequiredDuringExecution` reserved
+- Remove the node label 
+- Verify that the pod is not evicted
+- Add the node's label back 
+
+- Re-start kube-controller-manager, kube-scheduler and kube-apiserver with feature gate enabled
+- Remove the node label
+- Verify that the daemonSet pod on that node is evicted and not recreated.
+
+###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?
+
+No.
+
+### Monitoring Requirements
+
+###### How can an operator determine if the feature is in use by workloads?
+
+- Check the corresponding metrics(details in SLIs part below).
+- Inspect the `.spec.affinity.nodeAffinity.requiredDuringSchedulingRequiredDuringExecution` configuration.
+
+###### How can someone using this feature know that it is working for their instance?
+
+- [x] Events
+  - Event Reason: NodeAffinityEviction
+- [x] API .status
+  - Condition reason: DeletionByNodeAffinityEvictionController 
```

Comment:
How about this:
- `RequiredNodeAffinityViolation`

or maybe some of these:
- `NodeLabelsChanged`
- `NodeAffinityReconciliation` (eg, kube-controller-manager wins leader election, finds a violation)

We should aim to leave room for this field to one day become mutable, even if right now we think this is unlikely.